src/app/clusters/time-format-localization-server/time-format-localization-server.cpp

```diff
@@ -0,0 +1,222 @@
+    {
+        for (auto type : supportedCalendarTypes)
+        {
+            if (static_cast<uint8_t>(type) == newType)
+            {
+                res = chip::Protocols::InteractionModel::Status::Success;
+                break;
+            }
+        }
+    }
+
+    if (res != chip::Protocols::InteractionModel::Status::Success)
+    {
+        ChipLogError(Zcl, "Try to write invalid Calendar Type: %d", newType);
+    }
+
+    return res;
+}
+
+chip::Protocols::InteractionModel::Status __attribute__((weak))
+emberAfPluginTimeFormatLocalizationOnUnhandledAttributeChange(chip::EndpointId EndpointId, EmberAfAttributeType attrType,
+                                                              uint16_t attrSize, uint8_t * attrValue)
+{
+    return chip::Protocols::InteractionModel::Status::Success;
+}
+
+chip::Protocols::InteractionModel::Status MatterTimeFormatLocalizationClusterServerPreAttributeChangedCallback(
+    const chip::app::ConcreteAttributePath & attributePath, EmberAfAttributeType attributeType, uint16_t size, uint8_t * value)
+{
+    chip::Protocols::InteractionModel::Status res;
+
+    switch (attributePath.mAttributeId)
+    {
+    case chip::app::Clusters::TimeFormatLocalization::Attributes::HourFormat::Id:
+        if (sizeof(uint8_t) == size)
+        {
+            res = emberAfPluginTimeFormatLocalizationOnHourFormatChange(attributePath.mEndpointId, *value);
+        }
+        else
+        {
+            res = chip::Protocols::InteractionModel::Status::InvalidValue;
+        }
+        break;
+
+    case chip::app::Clusters::TimeFormatLocalization::Attributes::CalendarType::Id:
+        if (sizeof(uint8_t) == size)
+        {
+            res = emberAfPluginTimeFormatLocalizationOnCalendarTypeChange(attributePath.mEndpointId, *value);
+        }
+        else
+        {
+            res = chip::Protocols::InteractionModel::Status::InvalidValue;
+        }
+        break;
+
+    default:
+        res = emberAfPluginTimeFormatLocalizationOnUnhandledAttributeChange(attributePath.mEndpointId, attributeType, size, value);
+        break;
+    }
+
+    return res;
+}
+
+void emberAfTimeFormatLocalizationClusterServerInitCallback(chip::EndpointId endpoint)
+{
+    uint8_t calendarType;
+    DeviceLayer::AttributeList<TimeFormatLocalization::CalendarType, DeviceLayer::kMaxCalendarTypes> supportedCalendarTypes;
+    EmberAfStatus status = TimeFormatLocalization::Attributes::CalendarType::Get(endpoint, &calendarType);
+    VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                   ChipLogError(Zcl, "Failed to read calendar type with error: 0x%02x", status));
+
+    // To avoid OTA update removes a previously supported (and used) value from the list, checking whether the current Calendar Type
+    // value is allowed.
+    if (DeviceLayer::PlatformMgr().GetSupportedCalendarTypes(supportedCalendarTypes) == CHIP_NO_ERROR)
```

Comment:
This "check whether it's in our list" bit should be a helper function so we don't have two copies of it, no?

Reply:
It is not just check 'whether it's in our list', one of them also need to pick an valid value to replace the invalid default value

Reply:
Sure.  So for example, we could have a function that takes as input the given type and returns the given type if valid, or a (different) valid type if input is not valid.  The caller can then compare return value to what it passed in to decide whether it had a valid type, and if not it has a valid value to work with.

Reply:
Yes, there are plenty of ways to do it if we think this make more sense.